src/sentry/snuba/metrics/extraction.py

```diff
@@ -49,6 +52,66 @@
 
 logger = logging.getLogger(__name__)
 
+
+# This helps us control the different spec versions
+# in order to migrate customers from invalid specs
+class SpecVersion(NamedTuple):
+    version: int
+    flags: set[str] = set()
+
+    def matches_flags_set(self, flags: set[str]) -> bool:
+        return flags == self.flags
+
+
+class OnDemandMetricSpecVersioning:
+    # The flags from the spec versions alter the generated metric in OnDemandMetricSpec
+    # Once we're ready to abandon a version we will coalesce the list into a single element
+    # When there's a single version we should not have any flags and we should change
+    # OnDemandMetricSpec accordingly
+    spec_versions = [
+        SpecVersion(0),
+        SpecVersion(1, {"use_updated_env_logic"}),
+    ]
+
+    feature_to_flags_map = {
+        "organizations:on-demand-query-with-new-env-logic": {"use_updated_env_logic"},
+    }
+
+    @classmethod
+    def get_spec_versions(cls: Any) -> Sequence[SpecVersion]:
+        """Get all spec versions."""
+        return cls.spec_versions
+
+    @classmethod
+    def get_default_spec_version(cls: Any) -> SpecVersion:
+        # I believe this should be the original behaviour rather than the new one,
+        # however, the current code defaults to the new `use_updated_env_logic: bool = True``
+        return cls.spec_versions[-1]
+
+    @classmethod
+    def get_query_spec_version(cls: Any, feature_flag: str, org_id: int) -> SpecVersion:
+        """Return spec version based on feature flag enabled for an organization."""
+        org = Organization.objects.get_from_cache(id=org_id)
+        flags_set = (
+            cls.feature_to_flags_map[feature_flag] if features.has(feature_flag, org) else []
+        )
+
+        return cls._find_spec_version(flags_set)
```

Comment:
The matching logic is a bit overengineered, in my view we could have just done something like:
```suggestion
    def get_query_spec_version(cls: Any, feature_flag: str, org_id: int) -> SpecVersion:
        """Return spec version based on feature flag enabled for an organization."""
        if features.has("organizations:on-demand-query-with-new-env-logic", org):
        	return cls.spec_versions[1]

        return cls.spec_versions[0]
```

The logic you implemented is interesting but it's overly complicated (for our use case) since we do not need any complicated matching, we just need an if for the feature flag and return the spec version. Another mixed solution is to just define the spec_index given a feature flag but having a FIFO logic for matching rules seems a bit too much. No strong opinions here though. We can keep on using this.

Reply:
My initial idea was to have a simple function which given an org, checks for some feature flags and just returns the version that should be used for querying, nothing more. What do you think?

Reply:
@iambriccardo This is a good suggestion. I will include it. Thank you!